### ISSUE TITLE
META ISSUE: Revisit how imageio provides and handles meta data

### ISSUE THREAD

**Issue by almarklein** [imageio/imageio#362]
*This is a meta issue about meta data - in this top post I try to keep track of the current status. If you want to comment, see if there already is an issue related to your question/comment, open a new issue, or comment below.*

## Related issues

* https://github.com/scikit-image/scikit-image/issues/2605 - basically the skimage equivalent of this issue

## Goals

* Provide the user with a small subset of meta data with standardized fields and values.
* Provide the user with the full original raw metadata (differs per format).
* Make it possible for the user to read both of the above without actually reading the image data (where the format allows it).
* Make it possible to provide both standardized and format-specific metadata when writing images.

## Questions

* What USE CASES are there that we must consider? Let us know if you have a use-case that we're not covering yet! (see #501)
* What fields / values should the standardized metadata have (see #382)?
* How to deal with different units in metadata values (part of #382)?
* Should imageio return `(image, meta)` instead of having the meta data as a property (#503)?
* Should imageio report RGB vs HSV, or always convert to gray/RGB/RGBA (which I think is more or less what happens)? (#504)
* What should the metadata dict look like (#502)?

## Current status of imageio

* The read functions return an `imageio.Image` object (subclass of `ndarray`) which has a `.meta` dict attached.
* We have `Reader.get_meta_data()` but it does not have standard elements, so is usually not very helpful (#263).
* The (generic) Writer class allows giving meta data explicitly, and also checks for the `.meta` attribute.
* The high level public API (e.g. `imwrite()`) does not have a way to provide meta data explicitly.

## The plan / tasks

* [x] Rename and better docs for the ~`Image`~ `Array` class (done in #375, solving #333) 
* [ ] Document that code that processes images should probably not change its behavior based on the meta data itself, but only from args/kwargs that the user may set using the meta data.
* [ ] Define and implement standardized field names: #382, #502
* [ ] Better support for getting and setting meta data: #383 
* [ ] Implement meta data standardized fields on reading and writing for all plugins ... (this is probably where most of the work will go).

**Comment by lschr** [comment]
One could also use an approach similar to soft-matter/slicerator:
- Add an attribute to `Image` that contains a list of attributes that should be copied after image processing, e.g.
```python
class Image:
    _propagate_attrs = ["meta"]
    ...
```
- Provide a function decorator that can be used on processing functions which copies all attributes listed in `_propagate_attrs` to the result arrays.
```python
def copy_attrs(func):
     def wrapped(input, *args, **kwargs):
        res = func(input, *args, **kwargs)
        for a in getattr(input, "_propagate_attrs", []):
            setattr(res, a, getattr(input, a, None))
        return res
    return wrapped

@copy_attrs
def do_something(arr):
    return numpy.zeros_like(arr)
```
(not sure if above code actually works, but I think you get the idea).

**Comment by almarklein** [comment]
I can see how such an approach would make it easier to work with a custom `Image` class. The thing is that `imageio` does not have any processing functions itself, and `skimage` got rid of using a custom class some years ago. 

The purpose of this issue is to find a way to make meta data more (generically) useful (e.g. by standardizing certain info), let it work well with whatever `skimage` will decide on in https://github.com/scikit-image/scikit-image/issues/2605, and along the way, if possible, drop the `Image` class too.

**Comment by jakirkham** [comment]
What sort of metadata can the builtin plugins currently provide (including stuff that might not be nicely exposed currently)?

**Comment by almarklein** [comment]
@jakirkham That's a good question. I think that it should be possible to inspect `shape`, `dtype` and `nchannels` for most plugins, though many plugins will need work to do this *before* the loading of the data. There may also be plugins for which it is not possible to read metadata before reading the image data. What other kinds of generic metadata could we think of?

**Comment by lschr** [comment]
I think most formats support some kind of description or comment tag, which would be very helpful to have access to. Also date and time are very common.

**Comment by lschr** [comment]
Additionally, colorspace (what each channel represents; RGB or CMYK …) information may be useful.

**Comment by jni** [comment]
I totally missed this! Hi @almarklein! =) (And hi @jakirkham! =) (@lschr we haven't met but now it feels rude not to say hi. =P :wave: )

Regarding what metadata you could include, for microscopy an absolutely essential one is the pixel spacing. Some microscopes also do tiling and include e.g. stage positioning information, which then aids in mosaicing/registration to get the full image. Sometimes the units of the image data are also useful, e.g. "photons".

Regarding what to do with it, it goes far beyond copying and every function does something different. For example, downsampling requires one to increase the pixel spacing by a corresponding factor. Scaling the image requires adjusting the units. Colorspace conversion requires changing the colorspace. etc.

I think I just talked myself out of handling all this in scikit-image. =P

As to what imageio does with the Image class, from scikit-image's perspective I am agnostic. You can unpack into (array, dict) using `return_meta`, or you can keep returning `Image` and we can unpack it on our end easily enough.

I do like the idea of standard metadata fields and having imageio provide them (rather than doing the conversion on skimage's end). I can do a bit of work here both in discussing standard field names, and in providing an appropriate conversion for some formats (notably FEI, for which I wrote the reader).

One question is how to deal with the *existing* metadata tags that may or may not map to our standard ones. I think we would want to keep them, while still doing the conversion. For example, the FEI format has two layers of annotations, and pixel spacing is under `['Scan']['PixelHeight']` and `['Scan']['PixelWidth']`. (And these are in meters, which is simultaneously sensible and silly. =P) How do we prevent namespace clashes in the metadata hierarchy? I see two options:

- return *two* metadata dictionaries, one "raw" and one standardised
- use a special key, e.g. `['imageio']`, and if that key exists, check whether it's there from saving a previous imageio standard set, in which case we just keep it, or whether it's someone else's 'imageio' tag, in which case...? This gets messy quickly.

Anyway, they're my thoughts for now. Happy to have some friends in this adventure. =D

**Comment by lschr** [comment]
Hey @jni ;-)

After all, I gather it seems best to have an extra dict for metadata. I was thinking, that instead of having something like `return_meta=True` and changing the return signature accordingly, simply introducing a `meta` parameter may also be an option. This can either be `None`, in which case nothing happens or a user-supplied dictionary to be updated, like
```python
md = {}
img = imread(..., meta=md)
```
Not sure if this has any advantage; just an idea.

**Comment by jni** [comment]
@lschr Being a functional programming proponent, I'm quite opposed to modifying input arguments, especially in public-facing functions. =)

**Comment by jni** [comment]
@almarklein 

> have a subdict for the raw data.

Like, as a monkey-patched attribute? That doesn't seem like something I would describe as "nice and clean"???

**Comment by jni** [comment]
(Overall though, I'm glad we are converging on vaguely the same space, I think!)

**Comment by almarklein** [comment]
@jni, haha no, as having a dict with the standardized data, which has one field called `raw` that contains the raw meta data. As in the other way around as what you proposed.

**Comment by jni** [comment]
aaaaaah! :man_facepalming: Got it. =) Yes that's nice and clean, I agree!

**Comment by almarklein** [comment]
Another thing to fix is #371: the meta data can become invalid when it's "used" during reading, e.g. when exif rotation is applied. Upon saving, when the original meta data is provided (which happens by default), the image is already rotated but *also* has the exif rotation flag set.

The solution would be for the reader to modify the meta data (e.g. remove exif rotation flag). In which case, should the dict obtained via `imageio.metaread()` return the actual meta data as in the file, or as it is when the image is read?

**Comment by almarklein** [comment]
I have edited the top post to list the "plan of action", and made issues for more specific tasks.

**Comment by xkortex** [comment]
I would highly recommend leaving some indication that the original file had some metadata, even if (_especially_ if) it is transformed during load. 

Real life use case: I have a dataset of images of assorted datatypes, among them jpgs with orientation flag. I have a corresponding set of png images which are masks showing key regions in the original images. Here's the kicker, the png masks are always the same orientation of the raw image array (i.e. ignoring the orientation tag). But I also need to turn upright the portrait images, to feed them to a face  detector, which is fairly rotation-variant.

**Comment by FirefoxMetzger** [comment]
This is my week of necromancing issues, so I would like to pick up this meta-data discussion again and make a new proposal.

We do two modifications to the current plugin API:

1. We add a new property (called `descriptor`) to deal with standardized metadata. 
2. We add a new kwarg called `remove_applied=False` to `get_meta`. 

```python
class Plugin:
    @property
    def descriptor(self) -> "ImageDescriptor":
        raise NotImplementedError

    def get_meta(self, *, remove_applied=False): -> Dict[str, Any]
        raise NotImplementedError
```


The new function returns a `ImageDescriptor` of the following form:

```python
@dataclass
class ImageDescriptor:
    shape: Tuple[int, ...]
    dtype: DTypeLike
    colorspace: str
    channel_axis: int

```

This descriptor is _very_ minimal, but (I think) incorporates all the information necessary to prepare a numpy-based CV pipeline.  Plugins are allowed to `raise NotImplementedError()` if this behavior is not supported. In this case, end-users will have to fall back to `get_meta()` and manual parsing or `read()` + magic (a-priori knowledge). Additionally, (where possible) plugins should try to provide this data while reading as little data as possible.

The new keyword `remove_applied=False` does what it says on the tin. By default (`False`) `get_meta` will return a dict with any metadata items that exist in the format and that the plugin is able to extract. If set to `True`, the plugin should either `raise NotImplementedError()` if it doesn't wish to provide this functionality or remove all items that the plugin would apply during a read. For example, Pillow applies exif_rotation flags, so it would not add it to the metadata dict if `remove_applied=True`. The contents of this dict is free-for-all and plugin+format dependent, i.e., what ever the format allows minus what ever the plugin doesn't understand.

This would cover all use-cases listed in #501 , be - as far as I can see - backwards compatible with the current API, and it is implementable in an incremental way. 

@almarklein Thoughts?

**Comment by almarklein** [comment]
> This would cover all use-cases listed in #501 , be - as far as I can see - backwards compatible with the current API, and it is implementable in an incremental way.

That sounds pretty neat. Earlier I was thinking in the direction of a single dict with subdicts that contained *all the info*. But I like how the data that you get with this is much more concise. I think the descriptor is what some functions in skimage could accept and use to change their behavior, and not the (format-specific) metadata. I wonder if we an come up with a nicer name for "descriptor", let's chose a name that we'll use both in imageio and in skimage.

A few comments:

> The new function returns a ImageDescriptor

A dict would do, I think?

> This descriptor is very minimal, but (I think) incorporates all the information necessary to prepare a numpy-based CV pipeline. 

I think it would also need `spacing`, `origin` and `rotation`. If you compare/combine two datasets that cover a different region of a 2D/3D space.

> The new keyword `remove_applied=False` does what it says on the tin. By default (False)

I wonder if `remove_applied` should be True by default. I imagine a user reading an image + metadata, and the storing the imagedata +metadata again in the same format.

> Naming suggestions for these are welcome :)

Let's use verbs for functions: e.g `get_descriptor()`, `iio.read_metatdata()`, etc.

**Comment by FirefoxMetzger** [comment]
> I think the descriptor is what some functions in skimage could accept and use to change their behavior, and not the (format-specific) metadata. I wonder if we an come up with a nicer name for "descriptor", let's chose a name that we'll use both in imageio and in skimage.

Sure, I am happy to discuss naming. (As long as it doesn't delay us by another 2 years 🤣 . Who would be the best person to annoy about this in skimage? @jni has been involved in this discussion before, but maybe that is now someone else's job?

> > The new function returns a ImageDescriptor
>
> A dict would do, I think?

I would actually make it a class (could be a dataclass, but doesn't have to be), because of the better ergonomics. 

Type hinting and validation are really becoming a thing, and it is much easier to get with a class. There is also the (minor) advantage of `desc.shape` looking better and more familiar than `desc["shape"]`. 

It also has the advantage of enforcing certain content. Plugin authors would get an informative error when they forget an item, and they also can't sneak other items in, which would undermine the *standardized* part of the "standardized minimal set of metadata" 

> I think it would also need spacing, origin and rotation. If you compare/combine two datasets that cover a different region of a 2D/3D space.

I would really like to keep this minimal. If we add `spacing`, `origin`, and `rotation`, where do we draw the line from other relevant, but specialist elements like `fps` (for video), `location` (geotag), `aperture`, or `white balance`, etc.? We definitely need to provide all of this information, so it will definitely be in the `metadata` dict.

Currently, I'm actually debating the opposite: Should we toss out `colorspace`? For `shape`, `dtype`, `channel_axis` I can motivate direct performance implications, which makes them somewhat critical.  For `colorspace` (and the other fields we have discussed here) I only see them influencing the path a program/pipeline takes (i.e. they cause branching on the python level), but I don't see them having a direct impact on pretty much every low-level op or kernel in the pipeline. (Not sure if I managed to explain what I mean here.)


> I wonder if remove_applied should be True by default. I imagine a user reading an image + metadata, and the storing the imagedata +metadata again in the same format.

Yep, that would be nice indeed. However, I think it would break backward compatibility; at least with our pillow plugin. At the moment plugins do a metadata free-for-all, which usually means "return everything you can find". So - at least until v3 - this should stay the default, I think.

At the same time, nothing stops us from switching this around in V3 😃

> Let's use verbs for functions: e.g get_descriptor(), iio.read_metatdata(), etc.

I was thinking along similar lines, and `read_metadata` already has the name `get_meta` in the plugin API. Then I got reminded that our current items follow a `im<X>` pattern: `imread`, `imopen`, `imiter`, ... which aren't verbs. However, `immeta` doesn't work too nicely, nor does `imeta` (apple + facebook lolz), so that theme seems hard to extend. I like brevity so I ended up suggesting `metadata` and `descriptor`.

Maybe we could use `iio.describe` or `iio.imdesc`? Still not amazing, but +1 for less typing.

**Comment by almarklein** [comment]
> I would actually make it a class (could be a dataclass, but doesn't have to be), because of the better ergonomics.

Fair enough.

> I would really like to keep this minimal. If we add spacing, origin, and rotation, where do we draw the line from other relevant, but specialist elements like fps (for video), location (geotag), aperture, or white balance, etc.?

I think the line should be when the data could influence image processing functions. This is most definitely the case for spacing, so I veel strongly about including that. Maybe less so for origin and rotation. Their value could be instrumental in visualization, but I understand we need to draw a line *somewhere*.

> Currently, I'm actually debating the opposite: Should we toss out colorspace?

I think this is a different discussion. ATM imageio (implicitly assumes everything is RGB/RGBA. We could make that explicit and not deal with colorspace. Or we say that other colorspaces are allowed, and include that value in the descriptor.

Another thought: the descriptor has values that are always set, and values that are standardized but optional (usually because they apply to specific domains).

> > I wonder if remove_applied should be True by default.
>
> Yep, that would be nice indeed. However, I think it would break backward compatibility; 

I think one could argue that some plugins return metadata without removing/updating values that have been applied, can be considered a bug. Enough so to justify breaking compatibly here, IMO.

**Comment by FirefoxMetzger** [comment]
> I think one could argue that some plugins return metadata without removing/updating values that have been applied, can be considered a bug. Enough so to justify breaking compatibly here, IMO.

Done. Then we set `remove_applied=True` by default with the argument that Metadata hasn't been handled consistently before, and deviance from this behavior is considered a bug.

> [...] but I understand we need to draw a line somewhere.

I think this is the only open point now, and it has been the main blocker previously, too. Maybe the solution here could be to start small to get the ball rolling, and to then see what we else we can standardize for all plugins as we go down that road?

For `dtype` and `shape` I am aware of rather straightforward ways to get them from pillow, ffmpeg, and tifffile. These backends cover most of the commonly used formats, so I think the other backends/formats will fall in line. There are also explicit requests from different libraries (dask, scikit-image, PIMS) for these two fields, and I think you and I agree that these should be in there, too. This is a base to start from.

One option here could be to start off with `shape` and `dtype` and then see about the other fields as we go along. If we can guarantee their existence for commonly used formats, we include them step-by-step (one PR at a time).

> [a] [...] This is most definitely the case for spacing, so I veel strongly about including that.
> [b] the descriptor has values that are always set, and values that are standardized but optional

`spacing` is definitely useful. Does every format provide it as a metadata field? If not, what is the sensible thing to do if it is missing? I have the same question for `channel_axis` (what to do for grayscale/binary?), `colorspace`, `fps`, and the other fields we have mentioned so far.

I do see value in a standardized name for these fields though. Maybe we can merge the `descriptor` idea with the previous split of the metadata dict and use a tiered system?

1. `plugin.descriptor(...)` --> standard name + guaranteed existence with a sensible value (hence we can return a class/struct)
2. `plugin.metadata(..., standardized=True)` --> standard name, but may not exist (hence we return a dict)
2. `plugin.metadata(..., standardized=False)` --> format-specific name, and may not exist (hence we return a dict)

In this approach `spacing`, `fps`, `colorspace`, ..., can exist as standardized names that users can query for specialist needs. At the same time, we don't force a plugin to report any metadata beyond a minimal `ImageDescriptor` (shape/dtype), which reduces the burden on plugin authors. We also keep the option to retrieve metadata in an unopinionated way (`standardized=False`), since any standard name we choose will likely not always match the name a format might choose.

Most importantly, with this approach, we don't have to sort out these details now. We can see how useful each field is as time goes on, and then react to that by introducing additional standard names for commonly used fields, or by adding attributes to `descriptor` once we realize that we can sensibly support a name for all the formats we provide.

---

Also, my daily dose of naming suggestions. Most of them are probably junk, but maybe they will spark the right idea in someone.

**plugin.descriptor**:

- `iio.imspec` (potentially include the field `format:FileFormat` here)
- `iio.imheader`
- `iio.imID`

**plugin.get_meta**:

- `iio.get_meta`
- `iio.imdetails`
- `iio.improperties`
- `iio.imattributes`

**Comment by almarklein** [comment]
> Maybe the solution here could be to start small to get the ball rolling, and to then see what we else we can standardize for all plugins as we go down that road?

+1

>  Does every format provide it as a metadata field? If not, what is the sensible thing to do if it is missing?

No. The sensible default is `(1, 1)` for 2D and `(1,1,1)` for 3D. But I think it also makes sense that unset optional fields are `None`.

> I have the same question for channel_axis (what to do for grayscale/binary?)

Ha, good point. I guess `None` - there is not axis for channels.

What about something like this:
```py
class Descriptor:
    def __init__(self, *, shape, dtype, spacing=None, channel_axis=None, colorspace="rgb"):
        self.shape = shape
        self.dtype = dtype
        self.spacing = spacing
        self.channel_axis = channel_axis
        self.colorspace = colorspace
```

The required attributes don't have a default, so plugins are forced to provide these. The others are optional. We can also add more optional attributes here. Code that consumes the descriptor can check the optional attributes being `None`.

> `plugin.metadata(..., standardized=True)`

I don't see the need tor this if there already is `plugin.descriptor`. Let's stick to having one obvious way to do it. I think your idea to separate the standardized fields from the plugin-specific metadata (`plugin.descriptor` vs `plugin.get_meta`) is a good idea.

----

My 2ct: `iio.read_meta()`, `plugin.get_meta`.

Maybe for the name of what-we-now-call-descriptor we should drop an issue in skimage; they'll write that name much more in their codebase then we. Some suggestions to add "improps", "info". I've grown to like `descriptor` except for it being a a bit long.

I fail to remember: do we have have a plan for an API to read the image data + metadata / descriptor?

**Comment by FirefoxMetzger** [comment]
> No. The sensible default is (1, 1) for 2D and (1,1,1) for 3D. 

If `spacing=(1,) * img.ndim` is a sensible and expected default here, I am happy to add it to the descriptor. 

I'm also a bit worried about time series here, because spacing might not be trivial. What about a timelapse that uses a non-constant time axis? For example, microscopy pictures may be taken at 5 min, 30 min, 1 hour, 2 hours, 5 hours, 12 hours, and 24 hours, and may then be stacked into a TIFF. What is the expected value of `descriptor.spacing` in this case?

Assuming we set the spacing wrong, e.g., because we set it to the default, but the image actually wasn't recorded using that spacing. Would this problem be easy-to-detect or would it create hard-to-track/difficult-to-solve bugs? It's also one of the reservations I have about `colorspace`. If the descriptor specifies the wrong colorspace, how much harm would that do, and how easily can we detect that?

> But I think it also makes sense that unset optional fields are None.

Hmm, something makes me unhappy about this idea. There is nothing wrong with `None` having specific meaning (e.g., `channel_axis=None` means there is no channel axis in this image), but I am concerned about `None` for values that we don't know or that are missing. It seems very much like a slippery slope into "well we can also add `my_niche_field` here because yes it is only used by these two formats, but it can just be `None` for all others, so that's okay". A `dict` with a
`__missing__` method that returns `None` would be the clean implementation of this idea, and suddenly there is only a marginal difference between the descriptor and metadata `dict`.

I prefer the descriptor to be a collection of things we can sensibly define for _all_ formats. "You can rely on the existence and meaningfulness of this field" is much stronger a statement than "this field will be populated if the format has some metadata that serves the same or a similar purpose and if the plugin supports extracting it".

> I don't see the need tor this if there already is plugin.descriptor. Let's stick to having one obvious way to do it. I think your idea to separate the standardized fields from the plugin-specific metadata (plugin.descriptor vs plugin.get_meta) is a good idea.

Okay. Then we have to clarify what the descriptor should contain. Values for which we can guarantee existence, or standardized names that may get populated if possible.

---

> Maybe for the name of what-we-now-call-descriptor we should drop an issue in skimage

I will create an issue at skimage for the naming, once we agree on what the descriptor does 😃 . 

For `plugin.get_meta` I would actually go with `iio.get_meta` to match the names of the two (more consistent API naming is always good). Alternatively, we can change the name of both plugin method and API function but keep them close. Just like we did for `read`/`imread` or `write`/`imwrite`.

Maybe we can have `plugin.props`/`iio.improps` or `plugin.details`/`iio.imdetails` and have that return an `ImageDescriptor` object. That will keep the name around, allow for a catchy name for the API function, and allow for things like `descriptor = iio.improps(uri)`.

**Comment by almarklein** [comment]
> I'm also a bit worried about time series here, because spacing might not be trivial.

IMO time series data should be read with mimread or mvolread, so you get a list of arrays instead of a single packed array. I have found that to work much better during my PhD at least. But if there is a case where multiple frames are stacked into a single array, then the spacing very likely does not matter for the algorithms that you apply to it. Because if it does, it would imply that the algorithm assumes a spatial relation where there is none.

> Assuming we set the spacing wrong, e.g., because we set it to the default, but the image actually wasn't recorded using that spacing. 

Well at least we tried. Not setting the spacing would be just as wrong. I don't see your point, maybe I'm missing something?

> It seems very much like a slippery slope into "well we can also add my_niche_field here because yes it is only used by these two formats, but it can just be None for all others, so that's okay". [...] I prefer the descriptor to be a collection of things we can sensibly define for all formats.

 If we'd limit to stuff that all formats could provide, then you're only left with shape, and what's the point of that? E.g. I know that multiple functions accept a `spacing` arg right now (I wrote two of them).

I don't think it need to be a slippery slope, because we can apply a few relatively simple criteria:
* the field must say something about the pixels. E.g. no date, gps location.
* the field must not be specific to the acquisition of the image. E.g. no iso values, apertures, MRI settings, etc. because these are already in the meta dict, and any function using that info would be specific to that type of image.

**Comment by almarklein** [comment]
I was thinking about this, and I am not sure what this would look like:
```py
im = iio.imread(uri)  # ok
im, descriptor = ... ??
im, meta = ... ???
im, descriptor, meta = ... ???  -> we're not going to open a file 3 times, are we?
```

**Comment by FirefoxMetzger** [comment]
> Well at least we tried. Not setting the spacing would be just as wrong. I don't see your point, maybe I'm missing something?

Not setting spacing means the user has to search for that info elsewhere. If a user tries to use `descriptor.spacing` it will simply `raise` and let him/her know that this doesn't work. Setting the spacing wrong will produce a result, but that result is not correct. Since there won't be an exception more novice users will likely accept the result without questioning it, which - to me - is worse than seeing an exception.

> I don't think it need to be a slippery slope, because we can apply a few relatively simple criteria:

Hm, okay. Lets see if we can get this off the ground. Worst case, we can always depreciate values again if we realize that things don't work out or cause problems.

> I was thinking about this, and I am not sure what this would look like:

Right, I totally forgot to respond to this one. My bad. Files would be opened once, and it would look like this:

```python
with iio.imopen(uri) as file:
    descriptor = file.props(...)
    metadata = file.get_meta(...)
    img = file.read(...)
```

**Comment by jni** [comment]
Hello, zombie friends! 😂 🧟 Love rebooting this, thanks @FirefoxMetzger for kicking things off again!

My 2c on the above discussion:
- dataclass is good relative to dict. At napari we've been using pydantic Models for this with much success. I recommend it.
- spacing is 100% essential imho. Translation and rotation only slightly less so. For me, we should add these to the model and set them to None if they are not provided by the plugin. Other than channel_axis, I literally couldn't care less about the other descriptor items being discussed. 😂 By using lazy arrays (e.g. dask), we can include them directly in the image, making them redundant.
- uneven spacing can be handled by tuples-of-tuples, similar to dask chunks. `spacing=((1, 5, 15, 30), 5, 1, 1)`. I actually disagree with @almarklein that we shouldn't worry about it if it's uneven. But I don't think it's a showstopper.

I also think that waiting two years has been a good thing. The array community of practice has improved dramatically since we last discussed this and we are now at a stage where returning a duck-array that conforms to the [array API](https://data-apis.org/array-api/latest/API_specification/array_object.html) is a valid move. So, potentially we can go back to our own Image class that supports lazy indexing, metadata, `__array__` and `__array_function__`, etc?

Anyway, probably we need to set some time aside for a synchronous chat about all this. Can we do December? The next few weeks are pretty full-on for me.

**Comment by FirefoxMetzger** [comment]
> dataclass is good relative to dict. At napari we've been using pydantic Models for this with much success. I recommend it.

Yep, I rely on them as well in my own projects. The problem is that pypy isn't too happy about them (which is why they haven't made it into ImageIO yet).

> spacing is 100% essential imho. Translation and rotation only slightly less so.

I assume both you (@jni) and @almarklein are coming from a strong microscopy-focused background? That would at least explain why you two value the `spacing` field so much, whereas I don't (ML and robotics background).

> By using lazy arrays (e.g. dask), we can include them directly in the image, making them redundant.

I'm no expert, but I don't think it is that simple (https://github.com/dask/dask-image/issues/181). This metadata discussion sparked again precisely because dask could see significant performance gain if it can get `shape` and `dtype` without having to load pixel values. As far as I understand, that info is needed to set up the pipeline.

Part of the motivation for `ImageDescriptor` is to provide efficient access to such metadata without having to touch pixel data. I realize that this makes `ImageDescriptor` double-barreled (efficient metadata access vs. standard names for common fields). I proposed to split these two aspects (https://github.com/imageio/imageio/issues/362#issuecomment-968242215) but that doesn't seem to find a lot of resonance, so maybe we don't need it or there is a better solution that I am not seeing.

> So, potentially we can go back to our own Image class that supports lazy indexing, metadata, __array__ and __array_function__, etc?

Initially, I wanted to agree and say that it might be okay to have a metadata attribute considering how much numpy has moved to make `ArrayLike`s feel more at home. Then I realized that I much prefer an `out` arg (https://github.com/imageio/imageio/issues/670) that allows `read`/`imread` to fill an existing buffer/memoryview instead of always creating a new one. For that, we don't know if it is always valid to attach a metadata attribute.

That isn't to say that we should abandon the idea of returning an array subtype with lazy-loaded metadata attached. It is mainly to say that I would like to keep the primary way of obtaining metadata separate from the returned object. I'm happy to add a metadata attribute on top of this mechanism, e.g., when called with `out=None`.

> Anyway, probably we need to set some time aside for a synchronous chat about all this. Can we do December? The next few weeks are pretty full-on for me.

Sure, we can schedule that. If one of you makes a suggestion I'll see if I can accomodate.

**Comment by almarklein** [comment]
> I assume both you (@jni) and @almarklein are coming from a strong microscopy-focused background?

Medical imaging (e.g. CT, MRI). Anisotropic data is very common there. And being able to position multiple volumes correctly with respect to each other (origin, rotation) is also a common problem.


> uneven spacing can be handled by tuples-of-tuples, similar to dask chunks. 

What do you mean with uneven spacing? In my comments I meant that the spacing value for the time-dimension does not matter (because its not a spatial dimension). If multiple arrays in a timeseries have different spacing, then you should have a list of arrays rather than a single array. Or are you saying it must be possible to stack these in a single array, and use a tuples-of-tuples spacing to descibe the inner distances?

**Comment by almarklein** [comment]
> So, potentially we can go back to our own Image [...]

Mmm, I did not expect this. Not sure if I'm relieved or annoyed :)

> Can we do December? 

I'll probably have time in December.

**Comment by jni** [comment]
> I meant that the spacing value for the time-dimension does not matter (because its not a spatial dimension)

I think people working with time series data might disagree that spacing in time doesn't matter??? It is critical e.g. when estimating velocities. I also think that it's important from an API standpoint to have spacing match the number of dimensions, ie spacing[i] should be the spacing for array dimension i. When we can't (e.g. channels), use None?

> This metadata discussion sparked again precisely because dask could see significant performance gain if it can get shape and dtype without having to load pixel values. As far as I understand, that info is needed to set up the pipeline.

Yes, what I'm saying is something like:

```python
class Image:
    def __init__(filename):
        metadata = get_metadata_from_file(filename)
        self.shape = metadata.shape
        self.ndim = metadata.ndim
        self.dtype = metadata.dtype
        self.data = make_dask_array_from_file_and_meta(filename, metadata)
```

Then we add all of the array protocol methods and attributes.

>  Then I realized that I much prefer an out arg (#670) that allows read/imread to fill an existing buffer/memoryview instead of always creating a new one.

Ok, something like:

```python
class Image:
    def __init__(filename, out=None):
        metadata = get_metadata_from_file(filename)
        self.shape = metadata.shape
        self.ndim = metadata.ndim
        self.dtype = metadata.dtype
        self.metadata = metadata
        if out is not None:
            self._data = out
            self._init = False
        else:
            self._data = make_dask_array_from_file_and_meta(filename, metadata)
            self._init = True

    @property
    def data(self):
        if not self._init:
            read_to_buffer(filename, metadata, self._data)
        return self._data
```
?

> That would at least explain why you two value the spacing field so much, whereas I don't

Lucky guess? :joy: Remote sensing / satellite folks also like physical spacing info about their pixels.

> Sure, we can schedule that. If one of you makes a suggestion I'll see if I can accomodate.

Do you prefer evening or morning? With a 10h difference we'll have to be a bit creative, but it can be done. :blush:

**Comment by almarklein** [comment]
> I think people working with time series data might disagree that spacing in time doesn't matter???

Good point. I assumed that the unit of the values in `spacing` would be homogeneous. It did not occur to me that some could be spatial units (e.g. mm) and others could be temporal, or even something else. On that note, the channels could also have a relation that can be expressed with *some* spacing value!

**Comment by almarklein** [comment]
Thinking about this more ... I wonder whether we should distinguish here. If I call a skimage function that takes spacing into account, then it will *very* likely assume that this spacing is spatial. Should we somehow take precautions for this situation? Or is it up to the user to provide a slice that only includes spatial dimensions?

This brings me to another point:
```py
a_function_that_uses_spacing(im[1,...], descriptor???)
```

Looks like this brings us full circle to the custom Image class, so we can do this:
```py
>> im = iio.imread(...)
>> im.shape
(10, 512, 512)
>> im.descriptor.spacing
(1, 0.5, 0.5)
>> im2 = im[0, ...]
>> im2.shape
(512, 512)
>> im2.descriptor.spacing
(0.5, 0.5)
```

**Comment by FirefoxMetzger** [comment]
> Yes, what I'm saying is something like:
> [...]
> Then we add all of the array protocol methods and attributes.

@jni Ah, so you are proposing a different idea from what we have been discussing so far? This looks like `Image` aims to lazy-load pixels using Dask (ImageIO using Dask for performance improvement), whereas I am saying that Dask can use ImageIO for performance improvement. I might misunderstand your intention though.

Dask needs cheap access to `shape`+`dtype` information (what we currently call `ImageDescriptor`). Workers then load the actual pixel data eg. via `iio.imread`. From what I understand, however, this loading of pixel data currently happens twice. Once in the main thread to be immediately discarded after having extracted `shape`+`dtype`, and once again on the worker thread/node/... The first pixel-data read can often be avoided, and that is what I was trying to achieve with `ImageDescriptor`.

>Ok, something like:
> [...]
>?

I am even more convinced that we aren't currently not talking about the same thing 😅. My idea of `iio.imread(..., out=my_object)` is to take `my_object`, place the image data into it ("modify it in-place"), and to then return `my_object` instead of `ndarray`. Your suggestion looks like we would get `my_object`'s buffer, modify it (in-place) and return an `Image` that shares the buffer with `my_object`.

Essentially, what I want to be able to do with `out` is something like:

```python
# pinning memory is expensive
batch = torch.empty((10, 512, 512, 3), pin_memory=True)

for idx, img_path in enumerate(images.iterdir()):
    img = iio.imread(img_path, out=batch[idx, ...])
    # img is a torch.Tensor

# batch is populated and ready to be picked up by the GPU
```

> Lucky guess? 😂 Remote sensing / satellite folks also like physical spacing info about their pixels.

The community around skimage is surprisingly homogeneous 😛 and people doing hardware and/or real-time critical work usually think of python as too slow to even try (old habits die hard 🤷). This limits the number of viable guesses considerably 😆 

> Do you prefer evening or morning? With a 10h difference we'll have to be a bit creative, but it can be done. 😊

How morning is morning? I am in the gym between 6-8 am (CET) on weekdays. Depending on how long you think we need I can offer 5 am (CET), but I think @almarklein will not be too happy about that.

Otherwise, I think the evening (again CET) will be easier.

> If I call a skimage function that takes spacing into account, then it will very likely assume that this spacing is spatial. Should we somehow take precautions for this situation? Or is it up to the user to provide a slice that only includes spatial dimensions?
>
>Looks like this brings us full circle to the custom Image class, so we can do this:

I would be careful with functions that self-configure based on image metadata; it sounds a bit like this walks a line bordering magic. Being explicit about where additional metadata is being used will make the code easier to understand (albeit more verbose). It also makes things scalable, because functions don't have to rely on getting an `Image` as input, but can accept any `ArrayLike` combined with something that looks like a valid metadata struct.

**Comment by jni** [comment]
> Should we somehow take precautions for this situation? Or is it up to the user to provide a slice that only includes spatial dimensions?

Buyer beware, imho. :joy: 

> `a_function_that_uses_spacing(im[1,...], descriptor???)`

Yes, scikit-image will almost certainly continue to work with external parameters rather than an Image object. Here we would say eg `spacing=im[1, ...].spacing` as the second argument...?

> and to then return my_object instead of ndarray

This seems pointless? You already have my_object. Just do whatever you want to do with it once you've read into it? If you're worried that my `__init__` doesn't actually fill the buffer, that's fine, you can do `Image(filename, out=my_object).read()`?

> Your suggestion looks like we would get my_object's buffer, modify it (in-place) and return an Image that shares the buffer with my_object.

No, it doesn't share a buffer, it uses my_object directly as the _data. As long as my_object satisfies the Array API, that's ok.

At any rate, I'm pretty sure we can satisfy both your and my API requirements in a single API. Let's discuss more live. It'll be nice to meet you! :smiley: 

5am sounds painful even for me on the other end! :joy: I think two reasonable options are 8am Europe/6pm Australia (Or 8:30, dunno how much buffer you want after the gym @FirefoxMetzger!) or 10pm Europe/8am Australia. @almarklein?

**Comment by almarklein** [comment]
8am-8:30 is getting-kids-breakfast-and-to-school time here. 10pm would work for me. Is that doable @FirefoxMetzger ?

**Comment by almarklein** [comment]
> I would be careful with functions that self-configure based on image metadata; it sounds a bit like this walks a line bordering magic.

I have always been under the impression that one of the goals of this whole metadata exercise was that skimage functions can consume the standardized metadata that comes with an image.

**Comment by FirefoxMetzger** [comment]
@almarklein @jni 10pm CET / 8am ACT it is. Now for the more challenging part. Finding a date :D

> I have always been under the impression that one of the goals of this whole metadata exercise was that skimage functions can consume the standardized metadata that comes with an image.

Yep, I think on that part we are on the same page (though maybe I'd draw the circle broader than just around skimage). However, it is a different question on whether the metadata and the pixel data should be served together (struct/object) or individually. @jni seems to prefer the `Image` object route, whereas I think that it will be cleaner to split metadata and pixel data into separate functions.

**Comment by almarklein** [comment]
> though maybe I'd draw the circle broader than just around skimage

Definitely!

> However, it is a different question on whether the metadata and the pixel data should be served together (struct/object) or individually.

Right. I think it makes sense for functions to automatically use the standardized fields, so subsequently it makes sense for these to be attached to the image object (so that they travel along with the image). But for the plugin-specific metadata this would get muddy, and a function using that info might result in surprising (magical in a bad way) behaviour. I think this is a valid argument against putting the *specific* metadata on the Image.

**Comment by jni** [comment]
@FirefoxMetzger @almarklein can you discuss dates amongst yourselves? I generally don't start my day before 9am so all of my 8ams are currently free *except* Dec 16. :smiley: As long as it's a weekday!

Regarding the one vs two objects, the issue is that two objects is clunky when the two things belong together *and* a single object is strictly more general, because I can always do `result = filters.gaussian(np.asarray(image), sigma=5 / image.spacing)`.

As a general rule I am indeed *anti* custom-objects, and my preference here would actually be that NumPy/Dask provided a "metadata" dict/namespace for custom attributes, and we would just use that. But they don't, so we have to do something else. And, the other thing I've learned in my years of campaigning against objects, is that sometimes, if you're repeatedly writing `x1, y1, z1 = func(x0, y0, z0)`, then maybe you should bundle x, y, and z in an object.

As to whether skimage will take these custom objects as input... I guess it depends on how custom they are vs whether we can make them standard enough. Maybe we should think about xarray? But what I'd prefer is an "annotated array" library that meets our needs and maybe a few extra — e.g. one of the things I've always wanted from skimage is the ability to annotate whether an array is an image, some coordinates, or a segmentation, among others.

**Comment by FirefoxMetzger** [comment]
@jni @almarklein Awesome. I looked up both of your emails and sent a suggestion; seemed like the easier/faster thing to do.

> Regarding the one vs two objects, the issue is that two objects is clunky when the two things belong together and a single object is strictly more general

Doesn't that go both ways though? It's pretty easy to tear a single object into multiple bits, but it is just as possible to stack parts into one object downstream if downstream decides that it wants to consume images in this way:

```python
@dataclass
class Image:
    pixel_data: np.ndarray
    descriptor: ImageDescriptor
    metadata: Dict[str, Any]

with iio.imopen("some_file.png") as file:
    img = Image(file.read(), file.props(), file.metadata)

# how to elegantly set sigma=5*img.spacing?
# "sigma_multiplier" doesn't seem too intuitive
result = filters.gaussian(image, sigma_multiplier=5)
```

This touches upon another topic, which I think makes sense to talk about during the meeting: We have `iio.imopen` and the plugin interface which takes care of lazy for performance junkies. Maybe we can tweak that to serve your use-case @jni. 

A plugin instance is actually fairly close to a lazy-loaded `Image` object, and no matter our solution. It is missing numpy slicing (explicit choice in the new API https://github.com/imageio/imageio/issues/563#issuecomment-723508131) and it doesn't support `np.asarray(plugin_instance)` (files _may_ hold multiple images). You do get `plugin_instance.read() -> ndarray` and `plugin_instance.descriptor.spacing` though, so it has all the information you want/need in a single object.

> if you're repeatedly writing `x1, y1, z1 = func(x0, y0, z0)`, then maybe you should bundle x, y, and z in an object.

I agree. My angle here is that the metadata issue hasn't been resolved for years because it likely isn't as important to most users as is access to raw pixel data. So if the main use-case is `x1, _, _ = func(x0, y0, z0)` (aka. `np.asarray(iio.imread(...))`) then perhaps it is better to keep metadata separate (but easily accessible) for those who need it. 

At the same time, if you are just interested in metadata, pixel-data shouldn't slow you down. There are definite use-cases for that, e.g. to pre-configure a pipeline to then pump large quantities of them through efficiently (dask use-case). Something like a `iio.improps("my.tiff")` would fit that bill perfectly.

Then there is the situation where you need both (standardized) metadata _and_ pixel data. I think this is the one we are trying to solve now, right? The preferred solution is to get an `Image` object that contains both pixel data and metadata so that they stay connected and can be accessed easily. Returning an image object would have an impact on the two scenarios above, so my question here is if this happens often enough (compared to the other two scenarios) to motivate solving it like this.

The "skimage auto-config" feature falls into the last category, and I can see how this is desirable. At the same time, I agree with the Zen of python that "explicit is better than implicit", so if we do auto-config, I would like to explicitly pass in a `config` object so that (a) I know exactly which functions do auto-config, and (b) I get some means of introspecting and debugging what the auto-config does if I need that.

> e.g. one of the things I've always wanted from skimage is the ability to annotate whether an array is an image, some coordinates, or a segmentation, among others.

From an I/O perspective that would depend on the underlying format used. If the format supports exif tags or a comment field, we could easily write a `comment` into the metadata for those annotations. If not then those annotations would be lost once the image is saved, which is a bit sad and kind of defeats the point I guess.

**Comment by FirefoxMetzger** [comment]
It looks like we can soon bring this to a close. To summarize the meeting we had, we agreed on the following points:

- Plugins will get two new methods: `properties(self, index=None) -> ImageDescriptor` and `metadata(self, *, remove_applied=True) -> Dict[str, Any]`. Details are in https://github.com/imageio/imageio/issues/362#issuecomment-966630641
- The `ImageDescriptor` contains standardized metadata and will grow incrementally as we figure out what can be standardized. It will have the following fields to start with: `shape`, `dtype`, `spacing`
- The functional API will get two new convenience wrappers: `iio.improps` and `iio.imdetails`
- There will be a new object API that allows the creation of an `Image` object. This object is essentially a plugin, except that it allows slicing which triggers lazy-loading of pixel data.

@almarklein @jni Anything I missed?

**Comment by almarklein** [comment]
We also discussed an `ImageSequence` object next to the `Image` object, right?

Names ... 

Let's consistently call metadata "meta". So `iio.immeta` I guess?

I like "props" for the descriptor. Let's also make it an `ImageProps` object ;)

**Comment by FirefoxMetzger** [comment]
> We also discussed an ImageSequence object next to the Image object, right?

Yes. We should see how we can add this one. It can't have a direct slicing API, due to the multi-image formats that allow different shapes (and are non-stackable as a consequence). Maybe it can be a plugin that returns `Image` instead of `ndarray`? Then again, that could be handled by doing `plugin.read(like=iio.Image)`.

> Let's consistently call metadata "meta". So iio.immeta I guess?

That was my first thought but I think `iio.immeta` is quite prone to typos due to the `mm`. Not sure if we worry about that. The alternative could be to consistently call everything `details`: `iio.imdetails`, `plugin.details`

> I like "props" for the descriptor. Let's also make it an ImageProps object ;)

Yes, I was thinking the same actually. Didn't want to edit the previous post though, because of consistency. Lets go with `properties`, `improps`, and `ImageProperties`.
